In [2]:
!pip install ipython-sql pymysql
%load_ext sql
!pip install pymysql

In [3]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


Before beginning the analysis of the dataframe, the below code will add a new column titled 'product_revenue' which is the price of the product and purchase frequency multiplied together. The datatype will be of DECIMAL(20,2) since the column will be a monetary value. 

In [6]:
%%sql
ALTER TABLE sales
ADD COLUMN product_revenue DECIMAL(20, 2);
UPDATE sales
SET product_revenue = price * purchase_frequency;

 * mysql+pymysql://root:***@localhost/electronic_sales
(pymysql.err.OperationalError) (1060, "Duplicate column name 'product_revenue'")
[SQL: ALTER TABLE sales
ADD COLUMN product_revenue DECIMAL(20, 2);]
(Background on this error at: https://sqlalche.me/e/20/e3q8)


We have already connected to the database, so the following lines allow us to select all columns from the sales database in order to get a feel for the data. The output is kept clean by limiting to five products since there are about 9 thousand entries in the dataframe.

In [7]:
%%sql
SELECT * FROM sales
LIMIT 5;

 * mysql+pymysql://root:***@localhost/electronic_sales
5 rows affected.


[(5874, 'Smartphones', 'Other Brands', Decimal('312.95'), 18, 0, 2, 1, 0, Decimal('625.90')),
 (5875, 'Smart Watches', 'Samsung', Decimal('980.39'), 35, 1, 7, 2, 1, Decimal('6862.73')),
 (5876, 'Tablets', 'Samsung', Decimal('2606.72'), 63, 0, 1, 5, 1, Decimal('2606.72')),
 (5877, 'Smartphones', 'Samsung', Decimal('870.40'), 63, 1, 10, 3, 1, Decimal('8704.00')),
 (5878, 'Tablets', 'Sony', Decimal('1798.96'), 57, 0, 17, 3, 0, Decimal('30582.32'))]

We now see what the dataframe is like and around what values might be present. It is important to note that for the gender column a 0 corresponds to male purchases, while a 1 corresponds to women purchases. As with the purchase_intent column, a 0 means a no, that the customer age and gender demographic combination has minimum intent to purchase the certain item, and a 1 is the opposite result that there is a strong intent to buy the item. Now, let us see which brands are in the dataframe and how well they tend to perform. In combination with extracting the brands, we will select the total revenue for each brand  to see which companies earned the most money as well as their total sales. We can also select the average revenue per sale to see if any companies are underperforming due to just not making enough sales. 

In [8]:
%%sql
SELECT brand, 
       SUM(product_revenue) AS total_revenue,
       SUM(purchase_frequency) AS total_sales,
       ROUND((SUM(product_revenue) / SUM(purchase_frequency)), 2) AS revenue_per_sale
FROM sales
GROUP BY brand
ORDER BY total_revenue DESC;

 * mysql+pymysql://root:***@localhost/electronic_sales
5 rows affected.


[('Samsung', Decimal('29490957.36'), Decimal('18694'), Decimal('1577.56')),
 ('HP', Decimal('28042907.69'), Decimal('18195'), Decimal('1541.24')),
 ('Sony', Decimal('27574672.94'), Decimal('18334'), Decimal('1504.02')),
 ('Apple', Decimal('26840725.43'), Decimal('17505'), Decimal('1533.32')),
 ('Other Brands', Decimal('26649414.41'), Decimal('17764'), Decimal('1500.19'))]

Of the four brands, Samsung garnered the most total revenue from their products, while the collection of 'other brands' consisting of smaller companies took last place. What is also interesting is that although Apple ranks fourth for total revenue and last for total sales, their revenue per sale ranks third and is fairly close to HP which has the second highest revenue per sale. This tells us that Apple may be underperforming in their sales statistics and are not selling as much product that would maximize revenue, because if they were they would be higher up on the total sales and total revenue ranks that would more closely align with their revenue per sale ranking.

Another statistic in the dataframe, that of satisfaction, can be of note in seeing or explaining these brands' total revenues. One would intuitively think that brands with higher average satisfaction ratings from their customers would produce more revenue, as they may have the best product and consumers would recognize that and spend more with them. Below are the average satisfaction ratings from all sales along with each company's total revenue ranking.

In [9]:
%%sql
SELECT brand, 
       AVG(satisfaction) AS avg_satisfaction, 
       RANK() OVER(ORDER BY SUM(product_revenue) DESC) AS revenue_ranking
FROM sales
GROUP BY brand
ORDER BY avg_satisfaction DESC;

 * mysql+pymysql://root:***@localhost/electronic_sales
5 rows affected.


[('Apple', Decimal('3.0580'), 4),
 ('Other Brands', Decimal('3.0146'), 5),
 ('HP', Decimal('2.9813'), 2),
 ('Sony', Decimal('2.9659'), 3),
 ('Samsung', Decimal('2.9628'), 1)]

The truth is that higher satisfaction ratings for the brands in the dataframe do not directly lead to more revenue. For example, Apple and the 'other brands' category rank 1 and 2 in average satisfaction rating while they are respectively 4th and 5th for revenue. This communicates that there are far more factors related to producing revenue from products than just satisfaction ratings and strength of product. Next, we can further examine where the total revenue is coming from by breaking down what category of products generate what revenue totals.

There are only five categories of products being sold by each company, but we can see which products have the highest satisfaction, purchases, revenue, and revenue per sale to draw insights into which products companies should focus on selling to maximize earnings.

In [10]:
%%sql
SELECT category,
       AVG(satisfaction) AS avg_satisfaction,
       SUM(purchase_frequency) AS total_purchases,
       SUM(product_revenue) AS total_revenue,
       SUM(product_revenue) / SUM(purchase_frequency) AS revenue_per_sale,
       AVG(customer_age) AS avg_age
FROM sales
GROUP BY category
ORDER BY total_revenue DESC;

 * mysql+pymysql://root:***@localhost/electronic_sales
5 rows affected.


[('Laptops', Decimal('2.9951'), Decimal('18602'), Decimal('28337465.11'), Decimal('1523.355828'), Decimal('42.6216')),
 ('Headphones', Decimal('2.9994'), Decimal('18003'), Decimal('28074309.77'), Decimal('1559.423972'), Decimal('43.2100')),
 ('Smart Watches', Decimal('3.0017'), Decimal('17912'), Decimal('27779380.00'), Decimal('1550.880974'), Decimal('44.0464')),
 ('Tablets', Decimal('2.9864'), Decimal('17581'), Decimal('27220512.23'), Decimal('1548.291464'), Decimal('43.5342')),
 ('Smartphones', Decimal('2.9973'), Decimal('18394'), Decimal('27187010.72'), Decimal('1478.036899'), Decimal('43.3346'))]

We can see that laptops bring in the most sales and most total revenue, yet their satisfaction and revenue per sale numbers are low compared to the other products. This can mean that laptops are some of the most in demand products and will be bought regardless of if they have a lower satisfaction rating. Headphones, smart watches, and tablets are items that are not as necessary items for the common person as are laptops and smartphones, and they seem to bring in fairly good numbers and high revenues per sale that can be capitalized on if the demand for these products rises. Smartphones have the lowest revenue and revenue per sale despite having strong purchase numbers, but this may just be attributed to smartphones being more of a necessity than the other products which may drive the price down and limit revenue upside.

We can further break down the product purchase data by also grouping these categories with the brands that sell them. Below we can see how each company performs in sales with each of the product categories.

In [90]:
%%sql
SELECT brand,
	   category,
       SUM(purchase_frequency) AS total_sales,
       RANK() OVER(ORDER BY SUM(purchase_frequency) DESC) AS total_sales_rank,
       SUM(product_revenue) AS total_revenue,
       RANK() OVER(ORDER BY SUM(product_revenue) DESC) AS revenue_rank
FROM sales
GROUP BY brand, category
ORDER BY brand, category;

 * mysql+pymysql://root:***@localhost/electronic_sales
25 rows affected.


[('Apple', 'Headphones', Decimal('3463'), 20, Decimal('5400159.00'), 17),
 ('Apple', 'Laptops', Decimal('3464'), 19, Decimal('5215772.31'), 19),
 ('Apple', 'Smart Watches', Decimal('3398'), 24, Decimal('5108041.58'), 24),
 ('Apple', 'Smartphones', Decimal('3473'), 18, Decimal('5196027.00'), 20),
 ('Apple', 'Tablets', Decimal('3707'), 8, Decimal('5920725.54'), 3),
 ('HP', 'Headphones', Decimal('3612'), 14, Decimal('5869896.87'), 4),
 ('HP', 'Laptops', Decimal('3876'), 2, Decimal('5776699.17'), 7),
 ('HP', 'Smart Watches', Decimal('3614'), 13, Decimal('6045899.19'), 2),
 ('HP', 'Smartphones', Decimal('3686'), 10, Decimal('5484031.68'), 16),
 ('HP', 'Tablets', Decimal('3407'), 23, Decimal('4866380.78'), 25),
 ('Other Brands', 'Headphones', Decimal('3724'), 7, Decimal('5584811.82'), 11),
 ('Other Brands', 'Laptops', Decimal('3687'), 9, Decimal('5565315.14'), 12),
 ('Other Brands', 'Smart Watches', Decimal('3445'), 22, Decimal('5153748.36'), 23),
 ('Other Brands', 'Smartphones', Decimal('3559'), 16, Decimal('5186162.67'), 21),
 ('Other Brands', 'Tablets', Decimal('3349'), 25, Decimal('5159376.42'), 22),
 ('Samsung', 'Headphones', Decimal('3666'), 11, Decimal('5663253.18'), 9),
 ('Samsung', 'Laptops', Decimal('4010'), 1, Decimal('6420935.21'), 1),
 ('Samsung', 'Smart Watches', Decimal('3726'), 6, Decimal('5837289.09'), 5),
 ('Samsung', 'Smartphones', Decimal('3840'), 3, Decimal('5820447.73'), 6),
 ('Samsung', 'Tablets', Decimal('3452'), 21, Decimal('5749032.15'), 8),
 ('Sony', 'Headphones', Decimal('3538'), 17, Decimal('5556188.90'), 13),
 ('Sony', 'Laptops', Decimal('3565'), 15, Decimal('5358743.28'), 18),
 ('Sony', 'Smart Watches', Decimal('3729'), 5, Decimal('5634401.78'), 10),
 ('Sony', 'Smartphones', Decimal('3836'), 4, Decimal('5500341.64'), 15),
 ('Sony', 'Tablets', Decimal('3666'), 11, Decimal('5524997.34'), 14)]

From this we can see the breakdown of where these companies excel at and areas where they might have some extra work to do. As where Sony is consistently average across all categories in terms of revenue ranks and Samsung consistently performs pretty well over all categories, we can also see that Apple is stellar at their tablet sales, but below average in all other categories. We can also see that Apple just performs poorly in comparison to other brands in terms of total sales, while other brands such as Sony receive less for more sales and Samsung typically receives more for less sales. We can draw that Sony may not be doing a great job accurately pricing their products to maximize revenue, Samsung is currently experiencing a great business model, and Apple may need to start working on getting their name out.  

To better understand why these revenue numbers and rankings could possibly be what they are, we could also look at each products satisfaction rating for each company. Logically, a product with higher customer satisfaction ratings should perform the best in terms of revenue brought in. The below code shows the average satisfaction ratings per product and the satisfaction ranking, while also keeping the revenue ranking as a column.

In [47]:
%%sql 
SELECT brand,
	   category,
       AVG(satisfaction) AS avg_satisfaction,
       RANK() OVER(ORDER BY AVG(satisfaction) DESC) AS satisfaction_rank,
       RANK() OVER(ORDER BY SUM(product_revenue) DESC) AS revenue_rank
FROM sales
GROUP BY brand, category
ORDER BY brand, category;

 * mysql+pymysql://root:***@localhost/electronic_sales
25 rows affected.


[('Apple', 'Headphones', Decimal('3.1040'), 1, 17),
 ('Apple', 'Laptops', Decimal('3.0501'), 9, 19),
 ('Apple', 'Smart Watches', Decimal('3.0353'), 11, 24),
 ('Apple', 'Smartphones', Decimal('3.0933'), 2, 20),
 ('Apple', 'Tablets', Decimal('3.0108'), 13, 3),
 ('HP', 'Headphones', Decimal('3.0718'), 5, 4),
 ('HP', 'Laptops', Decimal('2.9370'), 19, 7),
 ('HP', 'Smart Watches', Decimal('3.0575'), 7, 2),
 ('HP', 'Smartphones', Decimal('2.9651'), 17, 16),
 ('HP', 'Tablets', Decimal('2.8785'), 23, 25),
 ('Other Brands', 'Headphones', Decimal('2.9889'), 14, 11),
 ('Other Brands', 'Laptops', Decimal('2.9780'), 16, 12),
 ('Other Brands', 'Smart Watches', Decimal('2.9888'), 15, 23),
 ('Other Brands', 'Smartphones', Decimal('3.0691'), 6, 21),
 ('Other Brands', 'Tablets', Decimal('3.0506'), 8, 22),
 ('Samsung', 'Headphones', Decimal('2.9091'), 22, 9),
 ('Samsung', 'Laptops', Decimal('3.0742'), 4, 1),
 ('Samsung', 'Smart Watches', Decimal('2.8496'), 24, 5),
 ('Samsung', 'Smartphones', Decimal('3.0207'), 12, 6),
 ('Samsung', 'Tablets', Decimal('2.9494'), 18, 8),
 ('Sony', 'Headphones', Decimal('2.9215'), 21, 13),
 ('Sony', 'Laptops', Decimal('2.9310'), 20, 18),
 ('Sony', 'Smart Watches', Decimal('3.0838'), 3, 10),
 ('Sony', 'Smartphones', Decimal('2.8488'), 25, 15),
 ('Sony', 'Tablets', Decimal('3.0456'), 10, 14)]

While most of the revenue rankings are fairly close to their satisfaction rankings, it is interesting to note Apple's interesting numbers. Despite being the only company with more than two average satisfaction ratings above 3, as they actually have all five, their revenue rankings for all products besides tablets is far lower than expected given their fantastic satisfaction ratings across the board. They, and perhaps the 'other brands' category, are underperforming for the most part given their satisfaction ratings while the other companies overperform their satisfaction ratings. We can also isolate the data to only get the brand and category with an average satisfaction rating of 3 to get a better look at the strong satisfaction performing categories

In [67]:
%%sql
SELECT brand, 
       category, 
       AVG(satisfaction) AS avg_satisfaction,
       RANK() OVER(ORDER BY AVG(satisfaction) DESC) AS satisfaction_rank
FROM sales
GROUP BY brand, category
HAVING AVG(satisfaction) > (
    SELECT AVG(satisfaction)
    FROM sales 
)
ORDER BY brand;

 * mysql+pymysql://root:***@localhost/electronic_sales
13 rows affected.


[('Apple', 'Headphones', Decimal('3.1040'), 1),
 ('Apple', 'Smartphones', Decimal('3.0933'), 2),
 ('Apple', 'Laptops', Decimal('3.0501'), 9),
 ('Apple', 'Smart Watches', Decimal('3.0353'), 11),
 ('Apple', 'Tablets', Decimal('3.0108'), 13),
 ('HP', 'Headphones', Decimal('3.0718'), 5),
 ('HP', 'Smart Watches', Decimal('3.0575'), 7),
 ('Other Brands', 'Smartphones', Decimal('3.0691'), 6),
 ('Other Brands', 'Tablets', Decimal('3.0506'), 8),
 ('Samsung', 'Laptops', Decimal('3.0742'), 4),
 ('Samsung', 'Smartphones', Decimal('3.0207'), 12),
 ('Sony', 'Smart Watches', Decimal('3.0838'), 3),
 ('Sony', 'Tablets', Decimal('3.0456'), 10)]

Yet as we know Apple has underperformed given their satisfaction ratings while some other companies have likely overperformed. Using a join operation and querying techniques we can display, for brand categories with an average satisfaction rating above 3, the actual total revenue and the estimated total revenue given the satisfaction rating. To do this, I calculated the revenue increase per satisfaction rating increase for the slope of the linear regression equation, and added this to the intercept, which was the average revenue over the entire data for products with a 0 satisfaction rating. I then fit the data and got this query and these results below.

In [52]:
%%sql
WITH revenue_per_satisfaction AS (
    SELECT 
        SUM(product_revenue) / SUM(satisfaction) AS revenue_per_satisfaction_unit
    FROM sales
)
SELECT brand, 
	   category, 
       AVG(satisfaction) AS avg_satisfaction,
       SUM(product_revenue) AS total_revenue,
       ROUND(((AVG(satisfaction) * 
       (SELECT revenue_per_satisfaction_unit
		FROM revenue_per_satisfaction)) + 
        (SELECT AVG(total_product_rev) - 
				(3 * (SELECT (SUM(product_revenue) / SUM(satisfaction)) AS rev_per_satisfaction_unit
					  FROM sales)) AS intercept
		 FROM (SELECT brand, 
					  category, 
					  SUM(product_revenue) AS total_product_rev, 
					  AVG(satisfaction) AS avg_satisfaction
			   FROM sales
			   GROUP BY brand, category) AS satisfaction_query)), 2) AS estimated_total_revenue
FROM sales
GROUP BY brand, category
HAVING AVG(satisfaction) > (
    SELECT AVG(satisfaction)
    FROM sales 
)
ORDER BY brand;

 * mysql+pymysql://root:***@localhost/electronic_sales
13 rows affected.


[('Apple', 'Headphones', Decimal('3.1040'), Decimal('5400159.00'), Decimal('5544481.93')),
 ('Apple', 'Laptops', Decimal('3.0501'), Decimal('5215772.31'), Decimal('5544204.84')),
 ('Apple', 'Smart Watches', Decimal('3.0353'), Decimal('5108041.58'), Decimal('5544128.53')),
 ('Apple', 'Smartphones', Decimal('3.0933'), Decimal('5196027.00'), Decimal('5544426.66')),
 ('Apple', 'Tablets', Decimal('3.0108'), Decimal('5920725.54'), Decimal('5544002.38')),
 ('HP', 'Headphones', Decimal('3.0718'), Decimal('5869896.87'), Decimal('5544316.38')),
 ('HP', 'Smart Watches', Decimal('3.0575'), Decimal('6045899.19'), Decimal('5544242.85')),
 ('Other Brands', 'Smartphones', Decimal('3.0691'), Decimal('5186162.67'), Decimal('5544302.10')),
 ('Other Brands', 'Tablets', Decimal('3.0506'), Decimal('5159376.42'), Decimal('5544207.18')),
 ('Samsung', 'Laptops', Decimal('3.0742'), Decimal('6420935.21'), Decimal('5544328.35')),
 ('Samsung', 'Smartphones', Decimal('3.0207'), Decimal('5820447.73'), Decimal('5544053.37')),
 ('Sony', 'Smart Watches', Decimal('3.0838'), Decimal('5634401.78'), Decimal('5544377.77')),
 ('Sony', 'Tablets', Decimal('3.0456'), Decimal('5524997.34'), Decimal('5544181.42'))]

Although it may be difficult to track, the higher the average satisfaction rating, the higher the estimated total revenue value (even if it is only marginally higher). What we can again draw from this is how Apple products and products from the 'Other Brands' category tend to perform below estimated values for their ratings while those of Samsung, HP, and Sony tend to perform at or above expectations. This can possibly be attributed to factors such as Apple and the 'Other Brands' not advertising as effectively as other companies or not pricing their products at the right price to maximize profit. Samsung for example has much lower satisfaction ratings than revenue ratings for the overall data for each category meaning that perhaps despite having an inferior overall product or buying experience, they are able to maximize their earnings. A way to look at maximizing earnings is by looking at target demographics for each product. Marketing and attempting to sell to people based on their age, gender, and the shopping tendencies that come with their unique demographic can improve a company's performance.

First let us break down the purchase tendencies by gender. The query below looks into the total sales by gender, spending per sale, the frequency of which each individual product is purchased, and the average intent for all products' given the gender.

In [91]:
%%sql
SELECT CASE WHEN gender = 1 THEN 'F' ELSE 'M' END AS gender,
	   SUM(purchase_frequency) AS total_sales,
       SUM(product_revenue) / SUM(purchase_frequency) AS spending_per_sale,
       AVG(purchase_frequency) AS avg_purchase_frequency,
       AVG(satisfaction) AS avg_satisfaction,
       AVG(purchase_intent) AS avg_intent
FROM sales
GROUP BY gender;

 * mysql+pymysql://root:***@localhost/electronic_sales
2 rows affected.


[('M', Decimal('44654'), Decimal('1525.989363'), Decimal('10.1027'), Decimal('2.9846'), Decimal('0.3120')),
 ('F', Decimal('45838'), Decimal('1537.090379'), Decimal('10.0083'), Decimal('3.0070'), Decimal('0.8120'))]

We can see that women do tend to make more purchases as well as spend more money per sale and item. It is however interesting to note that men do have a slightly higher purchase frequency per item, meaning that the items that they do buy, they tend to purchase more of as a collective whole. The largest difference in the table is the difference in average intent, with the females intent being over twice as high as the males'. This means that for the female population as a whole, they have a higher likeliness to buy products than their male counterparts by a large margin. 

Now that we have a slight grasp of the male and female spending split, we can further isolate the demographic metrics by breaking the gender category down in terms of age as well. The data contains the ages from 18 to 69, so there is a lot of data that may be overwhelming. Still, the key factors that may determine whether an age and gender group combination has signs for growth in the scope of marketing, or in companies pricing items that cater to purchase trends, are average purchase intent, average spending per item, and total spending. Another important factor in determining each demographic's purchase tendencies is their favorite category as well as their favorite brand. This will allow the companies that do not perform as well at different age and gender groups to market more so towards these underperforming groups, especially if these companies know their satisfaction ratings are higher than the competitors' in that category. Below is the query.

In [99]:
%%sql
WITH category_frequency AS (
    SELECT customer_age, 
           CASE WHEN gender = 1 THEN 'F' ELSE 'M' END AS gender,
           category, 
           SUM(purchase_frequency) AS purchase_count
    FROM sales
    GROUP BY customer_age, gender, category
),
brand_frequency AS (
    SELECT customer_age, 
           CASE WHEN gender = 1 THEN 'F' ELSE 'M' END AS gender,
           brand, 
           SUM(purchase_frequency) AS brand_count
    FROM sales
    GROUP BY customer_age, gender, brand
),
most_frequent_category AS (
    SELECT customer_age, gender, category,
           RANK() OVER (PARTITION BY customer_age, gender ORDER BY purchase_count DESC) AS category_rank
    FROM category_frequency
),
most_frequent_brand AS (
    SELECT customer_age, gender, brand, 
           RANK() OVER (PARTITION BY customer_age, gender ORDER BY brand_count DESC) AS brand_rank
    FROM brand_frequency
)
SELECT s.customer_age, 
       CASE WHEN s.gender = 1 THEN 'F' ELSE 'M' END AS gender, 
       AVG(s.purchase_intent) AS avg_intent,
       ROUND((SUM(product_revenue) / SUM(purchase_frequency)), 2) AS avg_spending,
       RANK() OVER(ORDER BY SUM(product_revenue) / SUM(purchase_frequency) DESC) AS avg_spending_rank,
       RANK() OVER(ORDER BY SUM(product_revenue) DESC) AS total_spending_rank,
       mfc.category AS most_frequent_category,
       mfb.brand AS most_frequent_brand
FROM sales s
LEFT JOIN most_frequent_category mfc 
    ON s.customer_age = mfc.customer_age 
   AND (CASE WHEN s.gender = 1 THEN 'F' ELSE 'M' END) = mfc.gender
   AND mfc.category_rank = 1  
LEFT JOIN most_frequent_brand mfb
    ON s.customer_age = mfb.customer_age
    AND (CASE WHEN s.gender = 1 THEN 'F' ELSE 'M' END) = mfb.gender
   AND mfb.brand_rank = 1 
GROUP BY s.customer_age, gender, mfc.category, mfb.brand
ORDER BY s.customer_age, gender;

 * mysql+pymysql://root:***@localhost/electronic_sales
105 rows affected.


[(18, 'F', Decimal('0.5714'), Decimal('1650.32'), 13, 8, 'Tablets', 'Sony'),
 (18, 'M', Decimal('0.0513'), Decimal('1308.58'), 104, 100, 'Tablets', 'Other Brands'),
 (19, 'F', Decimal('0.4468'), Decimal('1572.70'), 39, 25, 'Laptops', 'Sony'),
 (19, 'M', Decimal('0.0500'), Decimal('1407.40'), 93, 97, 'Laptops', 'Apple'),
 (20, 'F', Decimal('0.3816'), Decimal('1540.21'), 48, 88, 'Headphones', 'Other Brands'),
 (20, 'M', Decimal('0.0180'), Decimal('1538.02'), 53, 11, 'Laptops', 'Apple'),
 (21, 'F', Decimal('0.3514'), Decimal('1638.94'), 16, 1, 'Smartphones', 'HP'),
 (21, 'M', Decimal('0.0930'), Decimal('1371.66'), 98, 83, 'Smart Watches', 'Samsung'),
 (22, 'F', Decimal('0.3140'), Decimal('1635.59'), 17, 54, 'Headphones', 'Other Brands'),
 (22, 'M', Decimal('0.0694'), Decimal('1455.45'), 82, 103, 'Tablets', 'Samsung'),
 (23, 'F', Decimal('0.4019'), Decimal('1599.92'), 25, 2, 'Headphones', 'Samsung'),
 (23, 'M', Decimal('0.0227'), Decimal('1498.30'), 67, 86, 'Laptops', 'Apple'),
 (24, 'F', Decimal('0.3708'), Decimal('1595.40'), 27, 29, 'Laptops', 'HP'),
 (24, 'M', Decimal('0.0543'), Decimal('1418.53'), 89, 35, 'Laptops', 'Samsung'),
 (25, 'F', Decimal('0.3291'), Decimal('1751.06'), 4, 18, 'Headphones', 'Samsung'),
 (25, 'M', Decimal('0.0294'), Decimal('1590.06'), 31, 85, 'Laptops', 'HP'),
 (26, 'F', Decimal('0.5208'), Decimal('1592.61'), 28, 15, 'Smartphones', 'Samsung'),
 (26, 'M', Decimal('0.0353'), Decimal('1513.18'), 61, 34, 'Tablets', 'Samsung'),
 (27, 'F', Decimal('0.5000'), Decimal('1557.74'), 42, 47, 'Tablets', 'Samsung'),
 (27, 'M', Decimal('0.0357'), Decimal('1615.79'), 21, 17, 'Laptops', 'HP'),
 (28, 'F', Decimal('0.4022'), Decimal('1702.70'), 7, 7, 'Tablets', 'Other Brands'),
 (28, 'M', Decimal('0.0641'), Decimal('1592.24'), 29, 81, 'Headphones', 'Apple'),
 (29, 'F', Decimal('0.3667'), Decimal('1538.25'), 52, 49, 'Tablets', 'Samsung'),
 (29, 'M', Decimal('0.0595'), Decimal('1618.28'), 20, 42, 'Laptops', 'Other Brands'),
 (30, 'F', Decimal('0.3407'), Decimal('1577.01'), 35, 41, 'Smart Watches', 'Other Brands'),
 (30, 'M', Decimal('0.0417'), Decimal('1549.73'), 46, 14, 'Laptops', 'Sony'),
 (31, 'F', Decimal('0.9439'), Decimal('1573.28'), 38, 6, 'Laptops', 'Samsung'),
 (31, 'M', Decimal('0.3820'), Decimal('1504.99'), 65, 52, 'Laptops', 'Other Brands'),
 (32, 'F', Decimal('0.9762'), Decimal('1564.54'), 40, 82, 'Smart Watches', 'Other Brands'),
 (32, 'M', Decimal('0.4409'), Decimal('1610.65'), 22, 16, 'Smart Watches', 'Sony'),
 (33, 'F', Decimal('0.9740'), Decimal('1575.52'), 37, 101, 'Smartphones', 'Apple'),
 (33, 'M', Decimal('0.3942'), Decimal('1628.80'), 18, 4, 'Headphones', 'HP'),
 (34, 'F', Decimal('0.9529'), Decimal('1522.76'), 59, 39, 'Smartphones', 'Apple'),
 (34, 'M', Decimal('0.2742'), Decimal('1485.46'), 70, 105, 'Smartphones', 'HP'),
 (35, 'F', Decimal('0.9545'), Decimal('1408.25'), 92, 59, 'Smartphones', 'HP'),
 (35, 'M', Decimal('0.3816'), Decimal('1768.73'), 3, 57, 'Smartphones', 'Apple'),
 (36, 'F', Decimal('0.9506'), Decimal('1576.13'), 36, 67, 'Smartphones', 'Apple'),
 (36, 'M', Decimal('0.4321'), Decimal('1414.54'), 90, 96, 'Smartphones', 'Other Brands'),
 (37, 'F', Decimal('0.9759'), Decimal('1550.39'), 45, 75, 'Smartphones', 'Samsung'),
 (37, 'M', Decimal('0.4186'), Decimal('1482.60'), 72, 46, 'Smartphones', 'Sony'),
 (38, 'F', Decimal('0.9506'), Decimal('1724.73'), 6, 31, 'Laptops', 'Apple'),
 (38, 'M', Decimal('0.3793'), Decimal('1267.08'), 105, 89, 'Laptops', 'Sony'),
 (39, 'F', Decimal('0.9647'), Decimal('1460.18'), 78, 99, 'Smartphones', 'Sony'),
 (39, 'M', Decimal('0.3333'), Decimal('1590.16'), 30, 43, 'Headphones', 'Apple'),
 (40, 'F', Decimal('0.9195'), Decimal('1458.53'), 80, 44, 'Smartphones', 'Other Brands'),
 (40, 'M', Decimal('0.4783'), Decimal('1413.87'), 91, 45, 'Headphones', 'HP'),
 (41, 'F', Decimal('0.9882'), Decimal('1365.93'), 100, 95, 'Tablets', 'Apple'),
 (41, 'M', Decimal('0.6023'), Decimal('1384.61'), 95, 87, 'Laptops', 'Samsung'),
 (42, 'F', Decimal('0.9432'), Decimal('1487.15')

A lot of important insights can be drawn from this data. For example, while 25 year old females have the fourth highest average spending rank but they only have the 18th highest total spending rank, meaning that it is likely that if companies catered marketing towards them in perhaps the headphones department, they could see revenue increases that they had previously been missing out on as this group typically spends a lot but not in a lot of volume. This approach can be used for any group where the average spending rank is significantly higher than the total spending rank. Also in this same demographic group, we see that their favorite product is headphones and favorite brand is Samsung. Looking back at our earlier query with satisfaction ratings, we know that Samsung is actually last when it comes to headphone satisfaction ratings among all brands. So, a group like this, when marketed towards successfully by other brands with higher headphone satisfaction ratings, might sway consumers towards spending their money on the consensus superior products and brands. This can also lead to drawing in new fans and could perhaps contribute to the selling of more of the specific company's products. Average intent is of interest as well, as men have significantly lower intent than their female counterparts for every age, yet both genders' intent seems to grow as the age increases. Men also typically have higher spending averages according to the query, so although it is safer for companies to market to more closely guaranteed purchases like from women, if the brands attract males through their products and increase the male intent to purchase, the companies could harvest a new plethora of revenue previously not seen.

It is of note to see how these companies and their specific products have previously been marketed to each age and gender group. Displayed below is each brand and category grouped, but this time we are shown the average customer age and gender. As a reminder a 1 denotes a female purchase and a 0 denotes a male purchase, so an average of .55 would mean that the product is purchased by 55% females and 45% males. Also displayed is the count of products sold, the average satisfaction rating, average intent rating, and age ranking with 1 having the youngest customers.

In [81]:
%%sql
SELECT brand, 
       category, 
       AVG(customer_age) AS avg_customer_age, 
	   AVG(gender) AS avg_gender, 
       SUM(purchase_frequency) AS count_sold, 
	   AVG(satisfaction) AS avg_satisfaction,
       AVG(purchase_intent) AS avg_intent,
       RANK() OVER(ORDER BY AVG(customer_age)) AS age_rank
FROM sales
GROUP BY brand, category
ORDER BY brand, category;

 * mysql+pymysql://root:***@localhost/electronic_sales
25 rows affected.


[('Apple', 'Headphones', Decimal('43.1503'), Decimal('0.5347'), Decimal('3463'), Decimal('3.1040'), Decimal('0.5636'), 10),
 ('Apple', 'Laptops', Decimal('42.5460'), Decimal('0.5348'), Decimal('3464'), Decimal('3.0501'), Decimal('0.5933'), 6),
 ('Apple', 'Smart Watches', Decimal('44.8912'), Decimal('0.5647'), Decimal('3398'), Decimal('3.0353'), Decimal('0.6176'), 23),
 ('Apple', 'Smartphones', Decimal('43.3120'), Decimal('0.5598'), Decimal('3473'), Decimal('3.0933'), Decimal('0.6210'), 14),
 ('Apple', 'Tablets', Decimal('43.2204'), Decimal('0.5215'), Decimal('3707'), Decimal('3.0108'), Decimal('0.5430'), 12),
 ('HP', 'Headphones', Decimal('42.0603'), Decimal('0.5000'), Decimal('3612'), Decimal('3.0718'), Decimal('0.5862'), 2),
 ('HP', 'Laptops', Decimal('42.7559'), Decimal('0.4934'), Decimal('3876'), Decimal('2.9370'), Decimal('0.5643'), 7),
 ('HP', 'Smart Watches', Decimal('44.9260'), Decimal('0.5096'), Decimal('3614'), Decimal('3.0575'), Decimal('0.5781'), 24),
 ('HP', 'Smartphones', Decimal('42.1694'), Decimal('0.5054'), Decimal('3686'), Decimal('2.9651'), Decimal('0.5753'), 3),
 ('HP', 'Tablets', Decimal('44.9972'), Decimal('0.5198'), Decimal('3407'), Decimal('2.8785'), Decimal('0.5650'), 25),
 ('Other Brands', 'Headphones', Decimal('43.5014'), Decimal('0.5376'), Decimal('3724'), Decimal('2.9889'), Decimal('0.5766'), 16),
 ('Other Brands', 'Laptops', Decimal('43.2094'), Decimal('0.4518'), Decimal('3687'), Decimal('2.9780'), Decimal('0.5234'), 11),
 ('Other Brands', 'Smart Watches', Decimal('43.7978'), Decimal('0.5337'), Decimal('3445'), Decimal('2.9888'), Decimal('0.5843'), 19),
 ('Other Brands', 'Smartphones', Decimal('43.4006'), Decimal('0.4862'), Decimal('3559'), Decimal('3.0691'), Decimal('0.5608'), 15),
 ('Other Brands', 'Tablets', Decimal('44.3363'), Decimal('0.5030'), Decimal('3349'), Decimal('3.0506'), Decimal('0.5893'), 21),
 ('Samsung', 'Headphones', Decimal('44.1114'), Decimal('0.4868'), Decimal('3666'), Decimal('2.9091'), Decimal('0.5367'), 20),
 ('Samsung', 'Laptops', Decimal('42.4118'), Decimal('0.4936'), Decimal('4010'), Decimal('3.0742'), Decimal('0.5294'), 5),
 ('Samsung', 'Smart Watches', Decimal('43.6148'), Decimal('0.5172'), Decimal('3726'), Decimal('2.8496'), Decimal('0.5462'), 18),
 ('Samsung', 'Smartphones', Decimal('43.0646'), Decimal('0.5323'), Decimal('3840'), Decimal('3.0207'), Decimal('0.5995'), 8),
 ('Samsung', 'Tablets', Decimal('41.5927'), Decimal('0.4916'), Decimal('3452'), Decimal('2.9494'), Decimal('0.5056'), 1),
 ('Sony', 'Headphones', Decimal('43.2355'), Decimal('0.4971'), Decimal('3538'), Decimal('2.9215'), Decimal('0.5349'), 13),
 ('Sony', 'Laptops', Decimal('42.1753'), Decimal('0.4569'), Decimal('3565'), Decimal('2.9310'), Decimal('0.5259'), 4),
 ('Sony', 'Smart Watches', Decimal('43.0838'), Decimal('0.4973'), Decimal('3729'), Decimal('3.0838'), Decimal('0.5838'), 9),
 ('Sony', 'Smartphones', Decimal('44.7188'), Decimal('0.4934'), Decimal('3836'), Decimal('2.8488'), Decimal('0.5597'), 22),
 ('Sony', 'Tablets', Decimal('43.5926'), Decimal('0.5043'), Decimal('3666'), Decimal('3.0456'), Decimal('0.6040'), 17)]

In [94]:
%%sql
SELECT brand, 
       AVG(customer_age) AS avg_customer_age, 
	   AVG(gender) AS avg_gender, 
       SUM(purchase_frequency) AS count_sold, 
	   AVG(satisfaction) AS avg_satisfaction,
       AVG(purchase_intent) AS avg_intent,
       RANK() OVER(ORDER BY AVG(customer_age)) AS age_rank
FROM sales
GROUP BY brand
ORDER BY brand;

 * mysql+pymysql://root:***@localhost/electronic_sales
5 rows affected.


[('Apple', Decimal('43.4097'), Decimal('0.5426'), Decimal('17505'), Decimal('3.0580'), Decimal('0.5869'), 4),
 ('HP', Decimal('43.3742'), Decimal('0.5055'), Decimal('18195'), Decimal('2.9813'), Decimal('0.5736'), 2),
 ('Other Brands', Decimal('43.6385'), Decimal('0.5023'), Decimal('17764'), Decimal('3.0146'), Decimal('0.5664'), 5),
 ('Samsung', Decimal('42.9493'), Decimal('0.5049'), Decimal('18694'), Decimal('2.9628'), Decimal('0.5442'), 1),
 ('Sony', Decimal('43.3804'), Decimal('0.4899'), Decimal('18334'), Decimal('2.9659'), Decimal('0.5620'), 3)]

What we can see from this is that different companies tend to sell to different genders across their products as Apple heavily sells to mostly women, HP is rather neutral, Sony sells to mostly men, and Samsung and 'other brands' flip flop a decent margin depending on their product. What can also be seen is that because we know that women have on average a higher intent to purchase and are more satisfied, this can explain why these two numbers are so high for Apple. Apple seems to only be marketing and catering their products to women, which is why their ratings are so high but is also why their revenue numbers are lower than they should be. Also, our highest revenue reaping company Samsung, may possibly be performing so well due to their appeal with the young people and ability to market strongly product by product, regardless if the people they market towards may not be completely satisfied or have the highest intent to purchase. Overall, this breakdown lets us see the average ages and gender distribution for purchases so these companies can change their strategies and products to perhaps increase intent and satisfaction if their numbers are not what they desire. Yet, more importantly, this breakdown displays that companies should not be afraid to see a decrease in intent and satisfaction for the sake of appealing to a younger or male population that holds untapped revenue potential. 

While you may notice that all of the average ages are fairly similar in the table above, we can look at more age averages, but this time breaking the consumers down into age groups. The below groups are about as equal that I could get them, and each age group has displayed how many people are in the group, the average frequency of purchase, total revenue from each group, and the revenue per customer number and ranking. We can also see each groups favorite brand and category.

In [96]:
%%sql
WITH category_frequency AS (
    SELECT  
           CASE 
       WHEN customer_age BETWEEN 18 AND 24 THEN '18-24'
        WHEN customer_age BETWEEN 25 AND 31 THEN '25-31'
        WHEN customer_age BETWEEN 32 AND 39 THEN '32-39'
        WHEN customer_age BETWEEN 40 AND 46 THEN '40-46'
        WHEN customer_age BETWEEN 47 AND 54 THEN '47-54'
        WHEN customer_age BETWEEN 55 AND 61 THEN '55-61'
        WHEN customer_age BETWEEN 62 AND 69 THEN '62-69'
    END AS age_group,
           category, 
           SUM(purchase_frequency) AS purchase_count
    FROM sales
    GROUP BY age_group, category
),
brand_frequency AS (
    SELECT  
           CASE 
       WHEN customer_age BETWEEN 18 AND 24 THEN '18-24'
        WHEN customer_age BETWEEN 25 AND 31 THEN '25-31'
        WHEN customer_age BETWEEN 32 AND 39 THEN '32-39'
        WHEN customer_age BETWEEN 40 AND 46 THEN '40-46'
        WHEN customer_age BETWEEN 47 AND 54 THEN '47-54'
        WHEN customer_age BETWEEN 55 AND 61 THEN '55-61'
        WHEN customer_age BETWEEN 62 AND 69 THEN '62-69'
    END AS age_group,
           brand, 
           SUM(purchase_frequency) AS brand_count
    FROM sales
    GROUP BY age_group, brand
),
most_frequent_category AS (
    SELECT age_group, category,
           RANK() OVER (PARTITION BY age_group ORDER BY purchase_count DESC) AS category_rank
    FROM category_frequency
),
most_frequent_brand AS (
    SELECT age_group, brand, 
           RANK() OVER (PARTITION BY age_group ORDER BY brand_count DESC) AS brand_rank
    FROM brand_frequency
)
SELECT 
       CASE 
       WHEN s.customer_age BETWEEN 18 AND 24 THEN '18-24'
        WHEN s.customer_age BETWEEN 25 AND 31 THEN '25-31'
        WHEN s.customer_age BETWEEN 32 AND 39 THEN '32-39'
        WHEN s.customer_age BETWEEN 40 AND 46 THEN '40-46'
        WHEN s.customer_age BETWEEN 47 AND 54 THEN '47-54'
        WHEN s.customer_age BETWEEN 55 AND 61 THEN '55-61'
        WHEN s.customer_age BETWEEN 62 AND 69 THEN '62-69'
    END AS age_group,
        SUM(s.purchase_frequency) AS customers_per_group,
		AVG(s.satisfaction) AS avg_satisfaction,
		AVG(s.purchase_frequency) AS avg_frequency,
		SUM(s.product_revenue) AS total_rev,
		ROUND((SUM(s.product_revenue) / SUM(s.purchase_frequency)), 2) AS rev_per_customer,
		RANK() OVER(ORDER BY SUM(s.product_revenue) / SUM(s.purchase_frequency) DESC) AS rank_rev_per_customer,
        mfc.category AS most_frequent_category,
        mfb.brand AS most_frequent_brand
FROM sales s
LEFT JOIN most_frequent_category mfc 
    ON (CASE 
       WHEN s.customer_age BETWEEN 18 AND 24 THEN '18-24'
        WHEN s.customer_age BETWEEN 25 AND 31 THEN '25-31'
        WHEN s.customer_age BETWEEN 32 AND 39 THEN '32-39'
        WHEN s.customer_age BETWEEN 40 AND 46 THEN '40-46'
        WHEN s.customer_age BETWEEN 47 AND 54 THEN '47-54'
        WHEN s.customer_age BETWEEN 55 AND 61 THEN '55-61'
        WHEN s.customer_age BETWEEN 62 AND 69 THEN '62-69'
    END) = mfc.age_group
   AND mfc.category_rank = 1  
LEFT JOIN most_frequent_brand mfb
    ON (CASE 
       WHEN s.customer_age BETWEEN 18 AND 24 THEN '18-24'
        WHEN s.customer_age BETWEEN 25 AND 31 THEN '25-31'
        WHEN s.customer_age BETWEEN 32 AND 39 THEN '32-39'
        WHEN s.customer_age BETWEEN 40 AND 46 THEN '40-46'
        WHEN s.customer_age BETWEEN 47 AND 54 THEN '47-54'
        WHEN s.customer_age BETWEEN 55 AND 61 THEN '55-61'
        WHEN s.customer_age BETWEEN 62 AND 69 THEN '62-69'
    END) = mfb.age_group
   AND mfb.brand_rank = 1 
GROUP BY 
	CASE 
        WHEN s.customer_age BETWEEN 18 AND 24 THEN '18-24'
        WHEN s.customer_age BETWEEN 25 AND 31 THEN '25-31'
        WHEN s.customer_age BETWEEN 32 AND 39 THEN '32-39'
        WHEN s.customer_age BETWEEN 40 AND 46 THEN '40-46'
        WHEN s.customer_age BETWEEN 47 AND 54 THEN '47-54'
        WHEN s.customer_age BETWEEN 55 AND 61 THEN '55-61'
        WHEN s.customer_age BETWEEN 62 AND 69 THEN '62-69'
    END, mfc.category, mfb.brand
ORDER BY CASE 
        WHEN s.customer_age BETWEEN 18 AND 24 THEN '18-24'
        WHEN s.customer_age BETWEEN 25 AND 31 THEN '25-31'
        WHEN s.customer_age BETWEEN 32 AND 39 THEN '32-39'
        WHEN s.customer_age BETWEEN 40 AND 46 THEN '40-46'
        WHEN s.customer_age BETWEEN 47 AND 54 THEN '47-54'
        WHEN s.customer_age BETWEEN 55 AND 61 THEN '55-61'
        WHEN s.customer_age BETWEEN 62 AND 69 THEN '62-69'
    END;

 * mysql+pymysql://root:***@localhost/electronic_sales
7 rows affected.


[('18-24', Decimal('12570'), Decimal('2.9566'), Decimal('9.9132'), Decimal('19167677.04'), Decimal('1524.87'), 4, 'Laptops', 'Samsung'),
 ('25-31', Decimal('12641'), Decimal('3.0057'), Decimal('10.3024'), Decimal('20121007.27'), Decimal('1591.73'), 1, 'Laptops', 'Samsung'),
 ('32-39', Decimal('13154'), Decimal('2.9985'), Decimal('9.8606'), Decimal('20234243.73'), Decimal('1538.26'), 2, 'Smartphones', 'Sony'),
 ('40-46', Decimal('12220'), Decimal('3.0263'), Decimal('10.0411'), Decimal('18242068.39'), Decimal('1492.80'), 7, 'Laptops', 'Samsung'),
 ('47-54', Decimal('13788'), Decimal('2.9634'), Decimal('10.0937'), Decimal('21133160.50'), Decimal('1532.72'), 3, 'Smart Watches', 'Samsung'),
 ('55-61', Decimal('12301'), Decimal('2.9959'), Decimal('10.0416'), Decimal('18674139.89'), Decimal('1518.10'), 6, 'Smart Watches', 'HP'),
 ('62-69', Decimal('13818'), Decimal('3.0271'), Decimal('10.1379'), Decimal('21026381.01'), Decimal('1521.67'), 5, 'Smartphones', 'Sony')]

Most of the categories appear to have minimal satisfaction, frequency, and revenue differences across the age groups, but we can still see that revenue per customer appears to be the highest in the 25-31 and the 32-39 age ranges. This could mean that while older customers may bring in more revenue by sheer volume of old people buying products, these younger groups spend the most per person and may be provide the most room for future revenue growth. 

Lets break down the purchase and sales statistics by category of purchase now. To keep track of the total purchase ranking and revenue ranking and see how they compare between each age group and category combination.

In [97]:
%%sql
SELECT 
	   CASE 
       WHEN customer_age BETWEEN 18 AND 24 THEN '18-24'
        WHEN customer_age BETWEEN 25 AND 31 THEN '25-31'
        WHEN customer_age BETWEEN 32 AND 39 THEN '32-39'
        WHEN customer_age BETWEEN 40 AND 46 THEN '40-46'
        WHEN customer_age BETWEEN 47 AND 54 THEN '47-54'
        WHEN customer_age BETWEEN 55 AND 61 THEN '55-61'
        WHEN customer_age BETWEEN 62 AND 69 THEN '62-69'
    END AS age_group,
    category,
    SUM(purchase_frequency) AS sales_per_group,
	AVG(purchase_frequency) AS avg_frequency,
    SUM(product_revenue) AS total_rev,
    RANK() OVER(ORDER BY SUM(purchase_frequency) DESC) AS total_purchase_rank,
    RANK() OVER(ORDER BY (SUM(product_revenue) / SUM(purchase_frequency)) DESC) AS rev_per_purchase_rank
FROM sales
GROUP BY age_group, category
ORDER BY age_group, category;

 * mysql+pymysql://root:***@localhost/electronic_sales
35 rows affected.


[('18-24', 'Headphones', Decimal('2593'), Decimal('9.9349'), Decimal('4051428.22'), 16, 12),
 ('18-24', 'Laptops', Decimal('2755'), Decimal('9.9101'), Decimal('4311993.69'), 7, 11),
 ('18-24', 'Smart Watches', Decimal('2286'), Decimal('9.6050'), Decimal('3617000.15'), 32, 8),
 ('18-24', 'Smartphones', Decimal('2424'), Decimal('9.9344'), Decimal('3562882.59'), 26, 29),
 ('18-24', 'Tablets', Decimal('2512'), Decimal('10.1700'), Decimal('3624372.39'), 20, 31),
 ('25-31', 'Headphones', Decimal('2476'), Decimal('10.4915'), Decimal('4349572.23'), 25, 1),
 ('25-31', 'Laptops', Decimal('2665'), Decimal('10.4921'), Decimal('4137187.02'), 11, 14),
 ('25-31', 'Smart Watches', Decimal('2396'), Decimal('10.3723'), Decimal('3417333.74'), 29, 33),
 ('25-31', 'Smartphones', Decimal('2616'), Decimal('9.9468'), Decimal('3934753.70'), 15, 23),
 ('25-31', 'Tablets', Decimal('2488'), Decimal('10.2387'), Decimal('4282160.58'), 24, 2),
 ('32-39', 'Headphones', Decimal('2502'), Decimal('10.4250'), Decimal('3974751.47'), 21, 7),
 ('32-39', 'Laptops', Decimal('2725'), Decimal('9.5280'), Decimal('4369920.06'), 9, 4),
 ('32-39', 'Smart Watches', Decimal('2386'), Decimal('9.8595'), Decimal('3665117.72'), 30, 15),
 ('32-39', 'Smartphones', Decimal('3132'), Decimal('10.2353'), Decimal('4635686.34'), 2, 28),
 ('32-39', 'Tablets', Decimal('2409'), Decimal('9.2654'), Decimal('3588768.14'), 28, 27),
 ('40-46', 'Headphones', Decimal('2647'), Decimal('10.7602'), Decimal('3797104.07'), 12, 32),
 ('40-46', 'Laptops', Decimal('2682'), Decimal('10.5591'), Decimal('4047786.02'), 10, 21),
 ('40-46', 'Smart Watches', Decimal('2274'), Decimal('8.9528'), Decimal('3454749.97'), 33, 18),
 ('40-46', 'Smartphones', Decimal('2383'), Decimal('10.1404'), Decimal('3387096.19'), 31, 34),
 ('40-46', 'Tablets', Decimal('2234'), Decimal('9.7982'), Decimal('3555332.14'), 34, 6),
 ('47-54', 'Headphones', Decimal('2498'), Decimal('10.1959'), Decimal('3743216.08'), 22, 25),
 ('47-54', 'Laptops', Decimal('2859'), Decimal('10.1383'), Decimal('4320120.14'), 4, 20),
 ('47-54', 'Smart Watches', Decimal('3168'), Decimal('10.4211'), Decimal('4948817.38'), 1, 13),
 ('47-54', 'Smartphones', Decimal('2622'), Decimal('10.1236'), Decimal('3957227.69'), 14, 22),
 ('47-54', 'Tablets', Decimal('2641'), Decimal('9.5688'), Decimal('4163779.21'), 13, 9),
 ('55-61', 'Headphones', Decimal('2541'), Decimal('10.3714'), Decimal('3989434.49'), 18, 10),
 ('55-61', 'Laptops', Decimal('2423'), Decimal('10.0958'), Decimal('3355040.02'), 27, 35),
 ('55-61', 'Smart Watches', Decimal('2578'), Decimal('9.8397'), Decimal('4149086.71'), 17, 3),
 ('55-61', 'Smartphones', Decimal('2233'), Decimal('9.5837'), Decimal('3336189.44'), 35, 26),
 ('55-61', 'Tablets', Decimal('2526'), Decimal('10.3102'), Decimal('3844389.23'), 19, 17),
 ('62-69', 'Headphones', Decimal('2746'), Decimal('10.3623'), Decimal('4168803.21'), 8, 19),
 ('62-69', 'Laptops', Decimal('2493'), Decimal('10.0524'), Decimal('3795418.16'), 23, 16),
 ('62-69', 'Smart Watches', Decimal('2824'), Decimal('10.1219'), Decimal('4527274.33'), 5, 5),
 ('62-69', 'Smartphones', Decimal('2984'), Decimal('9.9136'), Decimal('4373174.77'), 3, 30),
 ('62-69', 'Tablets', Decimal('2771'), Decimal('10.2630'), Decimal('4161710.54'), 6, 24)]

It is interesting to see the breakdown of purchase and revenue rankings as some rows have similar values for both and some have very different values. This also ties back into the earlier query we looked into about the revenue disparity between different items. And although from the query right before the most recent one, the most commonly purchased items sometimes do not come fairly close to being at the top of the age groups' top revenue earner. This can tell the companies selling these products to be weary of towards chasing the groups that just buy the most volume, they should also look at which of each individual groups most bought product that also has a fairly high revenue per purchase rate. The 47-54 range brought in the most purchases out of any group and category in the laptop department, but this groups revenue per person was only 13th. Companies should make note of not of simply going to where the most volume and revenue is, but being smart and capitilizing where either the revenue per purchase is much lower than the total sales rank, to find areas that have the most upside, or find categories with the highest average frequencies that are not bringing in as much revenue. In these categories, we know that the consumers buy a lot of fewer products, meaning that if the right product is made with a high satisfaction rating for whatever company makes it, sales could go through the roof.

Instead of only looking at categorical breakdown, the code below shifts gears and looks at gender breakdown per group. The columns that are of note below are average satisfaction, average frequency, total revenue, and revenue per purchase as long as its rank per group.

In [100]:
%%sql
SELECT 
	   CASE 
       WHEN customer_age BETWEEN 18 AND 24 THEN '18-24'
        WHEN customer_age BETWEEN 25 AND 31 THEN '25-31'
        WHEN customer_age BETWEEN 32 AND 39 THEN '32-39'
        WHEN customer_age BETWEEN 40 AND 46 THEN '40-46'
        WHEN customer_age BETWEEN 47 AND 54 THEN '47-54'
        WHEN customer_age BETWEEN 55 AND 61 THEN '55-61'
        WHEN customer_age BETWEEN 62 AND 69 THEN '62-69'
    END AS age_group,
    CASE WHEN gender = 1 THEN 'F' ELSE 'M' END AS gender,
    SUM(purchase_frequency) AS customers_per_group,
    AVG(satisfaction) AS avg_satisfaction,
	AVG(purchase_frequency) AS avg_frequency,
    SUM(product_revenue) AS total_rev,
    ROUND((SUM(product_revenue) / SUM(purchase_frequency)), 2) AS rev_per_purchase,
    RANK() OVER(ORDER BY SUM(product_revenue) / SUM(purchase_frequency) DESC) AS rank_rev_per_purchase
FROM sales
GROUP BY age_group, gender
ORDER BY age_group, gender;

 * mysql+pymysql://root:***@localhost/electronic_sales
14 rows affected.


[('18-24', 'F', Decimal('6660'), Decimal('2.9879'), Decimal('10.0756'), Decimal('10704831.35'), Decimal('1607.33'), 2),
 ('18-24', 'M', Decimal('5910'), Decimal('2.9226'), Decimal('9.7364'), Decimal('8462845.69'), Decimal('1431.95'), 14),
 ('25-31', 'F', Decimal('6572'), Decimal('3.0233'), Decimal('10.2208'), Decimal('10605905.53'), Decimal('1613.80'), 1),
 ('25-31', 'M', Decimal('6069'), Decimal('2.9863'), Decimal('10.3921'), Decimal('9515101.74'), Decimal('1567.82'), 4),
 ('32-39', 'F', Decimal('6322'), Decimal('3.0437'), Decimal('9.5211'), Decimal('9770436.28'), Decimal('1545.47'), 7),
 ('32-39', 'M', Decimal('6832'), Decimal('2.9537'), Decimal('10.1970'), Decimal('10463807.45'), Decimal('1531.59'), 8),
 ('40-46', 'F', Decimal('6104'), Decimal('2.9168'), Decimal('9.9576'), Decimal('9007779.16'), Decimal('1475.72'), 12),
 ('40-46', 'M', Decimal('6116'), Decimal('3.1374'), Decimal('10.1258'), Decimal('9234289.23'), Decimal('1509.86'), 10),
 ('47-54', 'F', Decimal('6846'), Decimal('3.0281'), Decimal('10.1123'), Decimal('10644135.92'), Decimal('1554.80'), 5),
 ('47-54', 'M', Decimal('6942'), Decimal('2.8999'), Decimal('10.0755'), Decimal('10489024.58'), Decimal('1510.95'), 9),
 ('55-61', 'F', Decimal('6572'), Decimal('2.9567'), Decimal('10.1734'), Decimal('9803095.35'), Decimal('1491.65'), 11),
 ('55-61', 'M', Decimal('5729'), Decimal('3.0397'), Decimal('9.8946'), Decimal('8871044.54'), Decimal('1548.45'), 6),
 ('62-69', 'F', Decimal('6762'), Decimal('3.0828'), Decimal('10.0030'), Decimal('9920965.21'), Decimal('1467.16'), 13),
 ('62-69', 'M', Decimal('7056'), Decimal('2.9723'), Decimal('10.2707'), Decimal('11105415.80'), Decimal('1573.90'), 3)]

There is a fair spread of revenue per purchase coming from most of the groups, but the groups that should be targeted by companies should be those that have a strong combination of all factors. For example as with the 25-31 female group, they have the highest revenue per purchase, high overall total revenue, high satisfaction, and high frequency, meaning that they would be perfect for sustained and consistent sales. If I was Apple in the database and was lacking revenue, I would shift gears towards this younger audience of 25-31 year olds, both male and female, and sell them laptops and smartphones because that is what the group desires most, and that is what Apple sells best. 

The below query is fairly similar to the previous two displayed, yet is different in that the age groups are now split by brand.

In [78]:
%%sql
SELECT 
	   CASE 
       WHEN customer_age BETWEEN 18 AND 24 THEN '18-24'
        WHEN customer_age BETWEEN 25 AND 31 THEN '25-31'
        WHEN customer_age BETWEEN 32 AND 39 THEN '32-39'
        WHEN customer_age BETWEEN 40 AND 46 THEN '40-46'
        WHEN customer_age BETWEEN 47 AND 54 THEN '47-54'
        WHEN customer_age BETWEEN 55 AND 61 THEN '55-61'
        WHEN customer_age BETWEEN 62 AND 69 THEN '62-69'
    END AS age_group,
    brand,
    SUM(purchase_frequency) AS customers_per_group,
	AVG(purchase_frequency) AS avg_frequency,
    SUM(product_revenue) AS total_rev,
    RANK() OVER(ORDER BY (SUM(product_revenue) / SUM(purchase_frequency)) DESC) AS rev_rank
FROM sales
GROUP BY age_group, brand
ORDER BY age_group, brand;

 * mysql+pymysql://root:***@localhost/electronic_sales
35 rows affected.


[('18-24', 'Apple', Decimal('2485'), Decimal('9.7835'), Decimal('3797878.15'), 18),
 ('18-24', 'HP', Decimal('2508'), Decimal('9.6834'), Decimal('3978232.23'), 7),
 ('18-24', 'Other Brands', Decimal('2351'), Decimal('10.1336'), Decimal('3560925.51'), 23),
 ('18-24', 'Samsung', Decimal('2700'), Decimal('10.1504'), Decimal('4091294.94'), 22),
 ('18-24', 'Sony', Decimal('2526'), Decimal('9.8288'), Decimal('3739346.21'), 28),
 ('25-31', 'Apple', Decimal('2213'), Decimal('10.1514'), Decimal('3460164.81'), 10),
 ('25-31', 'HP', Decimal('2657'), Decimal('10.7571'), Decimal('4223124.63'), 6),
 ('25-31', 'Other Brands', Decimal('2652'), Decimal('10.2394'), Decimal('4097208.41'), 14),
 ('25-31', 'Samsung', Decimal('2751'), Decimal('10.2268'), Decimal('4476535.43'), 4),
 ('25-31', 'Sony', Decimal('2368'), Decimal('10.1197'), Decimal('3863973.99'), 3),
 ('32-39', 'Apple', Decimal('2685'), Decimal('9.5893'), Decimal('4237363.10'), 8),
 ('32-39', 'HP', Decimal('2387'), Decimal('9.5100'), Decimal('3631118.11'), 20),
 ('32-39', 'Other Brands', Decimal('2583'), Decimal('10.0506'), Decimal('3723998.01'), 33),
 ('32-39', 'Samsung', Decimal('2677'), Decimal('10.0262'), Decimal('4457371.20'), 1),
 ('32-39', 'Sony', Decimal('2822'), Decimal('10.1147'), Decimal('4184393.31'), 26),
 ('40-46', 'Apple', Decimal('2262'), Decimal('9.4644'), Decimal('3192814.63'), 35),
 ('40-46', 'HP', Decimal('2580'), Decimal('9.8473'), Decimal('3982383.61'), 15),
 ('40-46', 'Other Brands', Decimal('2336'), Decimal('9.8983'), Decimal('3366739.08'), 34),
 ('40-46', 'Samsung', Decimal('2588'), Decimal('10.1490'), Decimal('4133493.03'), 5),
 ('40-46', 'Sony', Decimal('2454'), Decimal('10.9067'), Decimal('3566638.04'), 30),
 ('47-54', 'Apple', Decimal('2777'), Decimal('10.3234'), Decimal('4302350.68'), 13),
 ('47-54', 'HP', Decimal('2717'), Decimal('9.9161'), Decimal('3946090.15'), 31),
 ('47-54', 'Other Brands', Decimal('2596'), Decimal('9.8707'), Decimal('4004800.83'), 16),
 ('47-54', 'Samsung', Decimal('2887'), Decimal('9.9552'), Decimal('4721003.70'), 2),
 ('47-54', 'Sony', Decimal('2811'), Decimal('10.4111'), Decimal('4158915.14'), 29),
 ('55-61', 'Apple', Decimal('2337'), Decimal('10.1169'), Decimal('3641955.88'), 11),
 ('55-61', 'HP', Decimal('2646'), Decimal('10.3359'), Decimal('4167162.29'), 9),
 ('55-61', 'Other Brands', Decimal('2539'), Decimal('9.9569'), Decimal('3685337.01'), 32),
 ('55-61', 'Samsung', Decimal('2340'), Decimal('9.7908'), Decimal('3535764.78'), 24),
 ('55-61', 'Sony', Decimal('2439'), Decimal('9.9959'), Decimal('3643919.93'), 25),
 ('62-69', 'Apple', Decimal('2746'), Decimal('10.2082'), Decimal('4208198.18'), 17),
 ('62-69', 'HP', Decimal('2700'), Decimal('9.9631'), Decimal('4114796.67'), 19),
 ('62-69', 'Other Brands', Decimal('2707'), Decimal('9.8796'), Decimal('4210405.56'), 12),
 ('62-69', 'Samsung', Decimal('2751'), Decimal('10.2649'), Decimal('4075494.28'), 27),
 ('62-69', 'Sony', Decimal('2914'), Decimal('10.3701'), Decimal('4417486.32'), 21)]

This further shows at what age groups companies are lacking within and could give insight to companies to either improve their standing with some age groups, if they are competitive there but losing to other brands, or pulling influence and focus on selling to other age groups. In the case where marketing may be pulled, this strategy may be more advantageous for the overall business strategy of the company.

The goal for companies to make more money is to make the best products possible and market them as effectively as they can towards the audiences with the best upside. And because we don't have any product analytics outside of just sales, we must have to continue to go off marketing towards the best groups of people. Another way to even more so narrow down the best groups to sell to is by finding the types of people with the best spending and frequency habits. We can look at which people that place in both the 80th and above percentiles for both of these categories to see who may be the very best to market towards. Below they are ranked by how much total money each group has spent.

In [102]:
%%sql 
WITH customer_spending AS (
    SELECT customer_age, 
           gender, 
           SUM(product_revenue) AS total_spent,
           AVG(purchase_frequency) AS avg_frequency
    FROM sales
    GROUP BY customer_age, gender
),
customer_rankings AS (
    SELECT customer_age, 
           gender, 
           total_spent, 
           avg_frequency,
           PERCENT_RANK() OVER (ORDER BY total_spent) AS spending_percentile,
           PERCENT_RANK() OVER (ORDER BY avg_frequency) AS frequency_percentile
    FROM customer_spending
)
SELECT customer_age, 
       CASE WHEN gender = 1 THEN 'F' ELSE 'M' END AS gender,
       total_spent,
       avg_frequency,
       spending_percentile,
       frequency_percentile
FROM customer_rankings
WHERE spending_percentile >= 0.8 AND frequency_percentile >= 0.8
ORDER BY total_spent DESC;

 * mysql+pymysql://root:***@localhost/electronic_sales
10 rows affected.


[(23, 'F', Decimal('1819112.07'), Decimal('10.6262'), 0.9902912621359223, 0.8446601941747572),
 (54, 'M', Decimal('1798232.96'), Decimal('11.4433'), 0.9805825242718447, 0.9902912621359223),
 (33, 'M', Decimal('1790047.37'), Decimal('10.5673'), 0.970873786407767, 0.8058252427184466),
 (28, 'F', Decimal('1731642.84'), Decimal('11.0543'), 0.941747572815534, 0.9514563106796117),
 (58, 'F', Decimal('1670134.71'), Decimal('10.8980'), 0.9223300970873787, 0.9223300970873787),
 (66, 'M', Decimal('1611371.42'), Decimal('11.5161'), 0.8932038834951457, 1.0),
 (57, 'F', Decimal('1596959.70'), Decimal('10.9010'), 0.883495145631068, 0.9320388349514563),
 (27, 'M', Decimal('1538231.17'), Decimal('11.3333'), 0.8446601941747572, 0.9805825242718447),
 (25, 'F', Decimal('1537427.35'), Decimal('11.1139'), 0.8349514563106796, 0.9611650485436893),
 (53, 'F', Decimal('1537420.66'), Decimal('10.6667'), 0.8252427184466019, 0.8543689320388349)]

These groups of people are ones that can be relied on to spend often and to spend big, so if a company is already excelling with these groups, they may want to branch out to other people as they should believe they can rely on these groups. On the contrary, if a company is not garnering enough earnings from these people, it would behoove them to market to these people. I also implemented an equation to help calculate who may be the best overall customers in terms of not just spending and frequency, but also in intent and satisfaction. The following are the order of the most overall valuable customers.

In [104]:
%%sql
WITH customer_lifetime_value AS (
    SELECT customer_age, 
           gender, 
           SUM(product_revenue) AS total_spent,
           ROUND(
            (SUM(product_revenue) / SUM(purchase_frequency)) * (AVG(purchase_frequency) * 5) * 
            (AVG(purchase_intent) / 1) * 
            (AVG(satisfaction) / 5),       
            2
        ) AS clv
    FROM sales
    GROUP BY customer_age, gender
)
SELECT customer_age, 
       CASE WHEN gender = 1 THEN 'F' ELSE 'M' END AS gender,
       total_spent,
       ROUND(clv, 2) AS lifetime_value
FROM customer_lifetime_value
ORDER BY clv DESC;

 * mysql+pymysql://root:***@localhost/electronic_sales
104 rows affected.


[(54, 'F', Decimal('1455326.38'), Decimal('53980.39')),
 (65, 'F', Decimal('1391284.03'), Decimal('50576.01')),
 (47, 'F', Decimal('1262760.06'), Decimal('49661.37')),
 (60, 'F', Decimal('1395927.71'), Decimal('49014.74')),
 (66, 'F', Decimal('1265314.18'), Decimal('48593.94')),
 (48, 'F', Decimal('1501734.11'), Decimal('48401.70')),
 (34, 'F', Decimal('1368960.02'), Decimal('47667.56')),
 (63, 'F', Decimal('1293298.25'), Decimal('47386.16')),
 (58, 'F', Decimal('1670134.71'), Decimal('47362.88')),
 (36, 'F', Decimal('1240416.86'), Decimal('47267.11')),
 (53, 'F', Decimal('1537420.66'), Decimal('47145.46')),
 (38, 'F', Decimal('1405651.64'), Decimal('46638.96')),
 (44, 'F', Decimal('1487200.92'), Decimal('46253.21')),
 (31, 'F', Decimal('1766798.34'), Decimal('46030.33')),
 (46, 'F', Decimal('1196228.15'), Decimal('45277.83')),
 (49, 'F', Decimal('1082335.74'), Decimal('45272.49')),
 (56, 'F', Decimal('1262958.81'), Decimal('44822.04')),
 (68, 'F', Decimal('1124006.42'), Decimal('44139.61')),
 (57, 'F', Decimal('1596959.70'), Decimal('43601.31')),
 (35, 'F', Decimal('1285732.86'), Decimal('43265.79')),
 (52, 'F', Decimal('1208962.71'), Decimal('43165.78')),
 (32, 'F', Decimal('1189053.07'), Decimal('43100.12')),
 (59, 'F', Decimal('1263912.66'), Decimal('42444.23')),
 (37, 'F', Decimal('1210857.51'), Decimal('41167.55')),
 (67, 'F', Decimal('1226641.74'), Decimal('41150.86')),
 (43, 'F', Decimal('1409429.95'), Decimal('40944.50')),
 (69, 'F', Decimal('1195653.78'), Decimal('40495.99')),
 (40, 'F', Decimal('1337473.57'), Decimal('40459.15')),
 (33, 'F', Decimal('1006754.40'), Decimal('40024.69')),
 (51, 'F', Decimal('1221102.65'), Decimal('39875.62')),
 (42, 'F', Decimal('1286383.42'), Decimal('39168.82')),
 (61, 'F', Decimal('1368659.35'), Decimal('37922.75')),
 (64, 'F', Decimal('1291993.11'), Decimal('37272.88')),
 (62, 'F', Decimal('1132773.70'), Decimal('37024.85')),
 (45, 'F', Decimal('1179197.47'), Decimal('36898.28')),
 (55, 'F', Decimal('1244542.41'), Decimal('36559.33')),
 (50, 'F', Decimal('1374493.61'), Decimal('36503.54')),
 (41, 'F', Decimal('1111865.68'), Decimal('36043.19')),
 (39, 'F', Decimal('1063009.92'), Decimal('35200.28')),
 (18, 'F', Decimal('1684974.56'), Decimal('32281.89')),
 (41, 'M', Decimal('1163069.50'), Decimal('28674.30')),
 (26, 'F', Decimal('1552791.58'), Decimal('28520.22')),
 (50, 'M', Decimal('1490207.83'), Decimal('26146.63')),
 (65, 'M', Decimal('1324979.88'), Decimal('25666.31')),
 (45, 'M', Decimal('1784784.09'), Decimal('24890.33')),
 (27, 'F', Decimal('1330311.58'), Decimal('24565.41')),
 (58, 'M', Decimal('1374946.55'), Decimal('23678.80')),
 (61, 'M', Decimal('1225235.46'), Decimal('23405.16')),
 (28, 'F', Decimal('1731642.84'), Decimal('23038.52')),
 (40, 'M', Decimal('1334691.48'), Decimal('22549.73')),
 (32, 'M', Decimal('1546226.83'), Decimal('22541.06')),
 (19, 'F', Decimal('1473621.46'), Decimal('22056.84')),
 (67, 'M', Decimal('1468407.73'), Decimal('22035.98')),
 (52, 'M', Decimal('1295930.95'), Decimal('21863.16')),
 (62, 'M', Decimal('1634459.06'), Decimal('21026.74')),
 (66, 'M', Decimal('1611371.42'), Decimal('20770.25')),
 (23, 'F', Decimal('1819112.07'), Decimal('20368.89')),
 (33, 'M', Decimal('1790047.37'), Decimal('20356.49')),
 (55, 'M', Decimal('1224845.42'), Decimal('19601.58')),
 (37, 'M', Decimal('1331374.09'), Decimal('19516.73')),
 (35, 'M', Decimal('1287635.27'), Decimal('19479.85')),
 (59, 'M', Decimal('1208212.78'), Decimal('18991.17')),
 (44, 'M', Decimal('1329928.15'), Decimal('18674.77')),
 (46, 'M', Decimal('1267210.29'), Decimal('18312.87')),
 (56, 'M', Decimal('1269018.68'), Decimal('18311.39')),
 (25, 'F', Decimal('1537427.35'), Decimal('18160.79')),
 (47, 'M', Decimal('1321118.90'), Decimal('18137.26')),
 (36, 'M', Decimal('1107581.14'), Decimal('17871.24')),
 (57, 'M', Decimal('1117980.62'), Decimal('17783.43')),
 (63, 'M', Decimal('1371436.16'), Decimal('17732.23')),
 (42, 'M', Decimal('1117728.55'), Decimal('17060.24')),
 (20, 'F', Decimal('1159

At the end of the day, the companies have to put far more into their scheme of operation than just by putting out a good product in order to churn revenue. A smart company markets to their strengths, and spots that can come potential strengths if they can outperform their competition in those facets. Marketing comes in many shapes and forms but is crucial towards the best companies earning the most money. Catering sales towards certain groups, designing certain products for specific people, and making note of the competitors numbers as well as strengths and weaknesses are all vital for a company who wishes to become the very best it can be. 

For example, we can previously note that Apple has the highest satisfaction rating for smartphones, yet they are outsold by other companies. We can note that one of the best groups to sell smartphones to are the 32-39 year olds, as that is their favorite product and that range in total has the second best revenue per sale. Yet, although we know Apple has the highest satisfaction of the group, HP and Sony come out on top for sales, as seen below.

In [11]:
%%sql
SELECT 
	   CASE 
       WHEN customer_age BETWEEN 18 AND 24 THEN '18-24'
        WHEN customer_age BETWEEN 25 AND 31 THEN '25-31'
        WHEN customer_age BETWEEN 32 AND 39 THEN '32-39'
        WHEN customer_age BETWEEN 40 AND 46 THEN '40-46'
        WHEN customer_age BETWEEN 47 AND 54 THEN '47-54'
        WHEN customer_age BETWEEN 55 AND 61 THEN '55-61'
        WHEN customer_age BETWEEN 62 AND 69 THEN '62-69'
    END AS age_group,
    brand,
    category,
    SUM(purchase_frequency) AS customers_per_group,
	AVG(purchase_frequency) AS avg_frequency,
    SUM(product_revenue) AS total_rev,
    RANK() OVER(ORDER BY (SUM(product_revenue) / SUM(purchase_frequency)) DESC) AS rev_per_sale_rank
FROM sales
WHERE CASE 
        WHEN customer_age BETWEEN 18 AND 24 THEN '18-24'
        WHEN customer_age BETWEEN 25 AND 31 THEN '25-31'
        WHEN customer_age BETWEEN 32 AND 39 THEN '32-39'
        WHEN customer_age BETWEEN 40 AND 46 THEN '40-46'
        WHEN customer_age BETWEEN 47 AND 54 THEN '47-54'
        WHEN customer_age BETWEEN 55 AND 61 THEN '55-61'
        WHEN customer_age BETWEEN 62 AND 69 THEN '62-69'
    END = '32-39' AND category = 'Smartphones'
GROUP BY age_group, brand, category
ORDER BY age_group, brand, category;

 * mysql+pymysql://root:***@localhost/electronic_sales
5 rows affected.


[('32-39', 'Apple', 'Smartphones', Decimal('603'), Decimal('10.3966'), Decimal('966516.75'), 1),
 ('32-39', 'HP', 'Smartphones', Decimal('731'), Decimal('10.1528'), Decimal('1053902.47'), 3),
 ('32-39', 'Other Brands', 'Smartphones', Decimal('637'), Decimal('10.4426'), Decimal('915765.81'), 4),
 ('32-39', 'Samsung', 'Smartphones', Decimal('499'), Decimal('10.1837'), Decimal('709172.86'), 5),
 ('32-39', 'Sony', 'Smartphones', Decimal('662'), Decimal('10.0303'), Decimal('990328.45'), 2)]

Now, we see that Apple actually still has the highest revenue per sale despite being third in total revenue for this group. Now, by taking the revenue per sale value and multiplying this value by the max customers per group in the table, which is 731, we can see how much each brand would sell if they had as many customers as HP in this demographic.

In [12]:
%%sql
SELECT 
	   CASE 
       WHEN customer_age BETWEEN 18 AND 24 THEN '18-24'
        WHEN customer_age BETWEEN 25 AND 31 THEN '25-31'
        WHEN customer_age BETWEEN 32 AND 39 THEN '32-39'
        WHEN customer_age BETWEEN 40 AND 46 THEN '40-46'
        WHEN customer_age BETWEEN 47 AND 54 THEN '47-54'
        WHEN customer_age BETWEEN 55 AND 61 THEN '55-61'
        WHEN customer_age BETWEEN 62 AND 69 THEN '62-69'
    END AS age_group,
    brand,
    category,
    SUM(purchase_frequency) AS customers_per_group,
	AVG(purchase_frequency) AS avg_frequency,
    SUM(product_revenue) AS total_rev,
    RANK() OVER(ORDER BY (SUM(product_revenue) / SUM(purchase_frequency)) DESC) AS rev_per_purchase_rank,
    SUM(product_revenue) / SUM(purchase_frequency) AS rev_per_purchase,
    (SUM(product_revenue) / SUM(purchase_frequency)) * (SELECT MAX(customers_per_group)
FROM 
(SELECT 
	   CASE 
       WHEN customer_age BETWEEN 18 AND 24 THEN '18-24'
        WHEN customer_age BETWEEN 25 AND 31 THEN '25-31'
        WHEN customer_age BETWEEN 32 AND 39 THEN '32-39'
        WHEN customer_age BETWEEN 40 AND 46 THEN '40-46'
        WHEN customer_age BETWEEN 47 AND 54 THEN '47-54'
        WHEN customer_age BETWEEN 55 AND 61 THEN '55-61'
        WHEN customer_age BETWEEN 62 AND 69 THEN '62-69'
    END AS age_group,
    brand,
    category,
    SUM(purchase_frequency) AS customers_per_group,
	AVG(purchase_frequency) AS avg_frequency,
    SUM(product_revenue) AS total_rev,
    RANK() OVER(ORDER BY (SUM(product_revenue) / SUM(purchase_frequency)) DESC) AS rev_rank
FROM sales
WHERE CASE 
        WHEN customer_age BETWEEN 18 AND 24 THEN '18-24'
        WHEN customer_age BETWEEN 25 AND 31 THEN '25-31'
        WHEN customer_age BETWEEN 32 AND 39 THEN '32-39'
        WHEN customer_age BETWEEN 40 AND 46 THEN '40-46'
        WHEN customer_age BETWEEN 47 AND 54 THEN '47-54'
        WHEN customer_age BETWEEN 55 AND 61 THEN '55-61'
        WHEN customer_age BETWEEN 62 AND 69 THEN '62-69'
    END = '32-39' AND category = 'Smartphones'
GROUP BY age_group, brand, category
ORDER BY age_group, brand, category) AS sub_query) AS would_be_total_rev
FROM sales
WHERE CASE 
        WHEN customer_age BETWEEN 18 AND 24 THEN '18-24'
        WHEN customer_age BETWEEN 25 AND 31 THEN '25-31'
        WHEN customer_age BETWEEN 32 AND 39 THEN '32-39'
        WHEN customer_age BETWEEN 40 AND 46 THEN '40-46'
        WHEN customer_age BETWEEN 47 AND 54 THEN '47-54'
        WHEN customer_age BETWEEN 55 AND 61 THEN '55-61'
        WHEN customer_age BETWEEN 62 AND 69 THEN '62-69'
    END = '32-39' AND category = 'Smartphones'
GROUP BY age_group, brand, category
ORDER BY age_group, brand, category;

 * mysql+pymysql://root:***@localhost/electronic_sales
5 rows affected.


[('32-39', 'Apple', 'Smartphones', Decimal('603'), Decimal('10.3966'), Decimal('966516.75'), 1, Decimal('1602.847015'), Decimal('1171681.167910')),
 ('32-39', 'HP', 'Smartphones', Decimal('731'), Decimal('10.1528'), Decimal('1053902.47'), 3, Decimal('1441.727045'), Decimal('1053902.470000')),
 ('32-39', 'Other Brands', 'Smartphones', Decimal('637'), Decimal('10.4426'), Decimal('915765.81'), 4, Decimal('1437.622936'), Decimal('1050902.365949')),
 ('32-39', 'Samsung', 'Smartphones', Decimal('499'), Decimal('10.1837'), Decimal('709172.86'), 5, Decimal('1421.188096'), Decimal('1038888.498316')),
 ('32-39', 'Sony', 'Smartphones', Decimal('662'), Decimal('10.0303'), Decimal('990328.45'), 2, Decimal('1495.964426'), Decimal('1093549.995392'))]

We can see that if Apple marketed significantly more and even tied HP in their sales in this category(even though we may know that according to satisfaction rating, Apple should have the ability to surpass them), Apple would receive a large improvement in revenue. To see the percentage of increase that is possible when Apple puts more resources into this category and age range, we can execute the following query that divides the would be total revenue by the actual total revenue.

In [13]:
%%sql
SELECT would_be_total_rev / total_rev
FROM
(SELECT 
	   CASE 
       WHEN customer_age BETWEEN 18 AND 24 THEN '18-24'
        WHEN customer_age BETWEEN 25 AND 31 THEN '25-31'
        WHEN customer_age BETWEEN 32 AND 39 THEN '32-39'
        WHEN customer_age BETWEEN 40 AND 46 THEN '40-46'
        WHEN customer_age BETWEEN 47 AND 54 THEN '47-54'
        WHEN customer_age BETWEEN 55 AND 61 THEN '55-61'
        WHEN customer_age BETWEEN 62 AND 69 THEN '62-69'
    END AS age_group,
    brand,
    category,
    SUM(purchase_frequency) AS customers_per_group,
	AVG(purchase_frequency) AS avg_frequency,
    SUM(product_revenue) AS total_rev,
    RANK() OVER(ORDER BY (SUM(product_revenue) / SUM(purchase_frequency)) DESC) AS rev_per_purchase_rank,
    SUM(product_revenue) / SUM(purchase_frequency) AS rev_per_purchase,
    (SUM(product_revenue) / SUM(purchase_frequency)) * (SELECT MAX(customers_per_group)
FROM 
(SELECT 
	   CASE 
       WHEN customer_age BETWEEN 18 AND 24 THEN '18-24'
        WHEN customer_age BETWEEN 25 AND 31 THEN '25-31'
        WHEN customer_age BETWEEN 32 AND 39 THEN '32-39'
        WHEN customer_age BETWEEN 40 AND 46 THEN '40-46'
        WHEN customer_age BETWEEN 47 AND 54 THEN '47-54'
        WHEN customer_age BETWEEN 55 AND 61 THEN '55-61'
        WHEN customer_age BETWEEN 62 AND 69 THEN '62-69'
    END AS age_group,
    brand,
    category,
    SUM(purchase_frequency) AS customers_per_group,
	AVG(purchase_frequency) AS avg_frequency,
    SUM(product_revenue) AS total_rev,
    RANK() OVER(ORDER BY (SUM(product_revenue) / SUM(purchase_frequency)) DESC) AS rev_rank
FROM sales
WHERE CASE 
        WHEN customer_age BETWEEN 18 AND 24 THEN '18-24'
        WHEN customer_age BETWEEN 25 AND 31 THEN '25-31'
        WHEN customer_age BETWEEN 32 AND 39 THEN '32-39'
        WHEN customer_age BETWEEN 40 AND 46 THEN '40-46'
        WHEN customer_age BETWEEN 47 AND 54 THEN '47-54'
        WHEN customer_age BETWEEN 55 AND 61 THEN '55-61'
        WHEN customer_age BETWEEN 62 AND 69 THEN '62-69'
    END = '32-39' AND category = 'Smartphones'
GROUP BY age_group, brand, category
ORDER BY age_group, brand, category) AS sub_query) AS would_be_total_rev
FROM sales
WHERE CASE 
        WHEN customer_age BETWEEN 18 AND 24 THEN '18-24'
        WHEN customer_age BETWEEN 25 AND 31 THEN '25-31'
        WHEN customer_age BETWEEN 32 AND 39 THEN '32-39'
        WHEN customer_age BETWEEN 40 AND 46 THEN '40-46'
        WHEN customer_age BETWEEN 47 AND 54 THEN '47-54'
        WHEN customer_age BETWEEN 55 AND 61 THEN '55-61'
        WHEN customer_age BETWEEN 62 AND 69 THEN '62-69'
    END = '32-39' AND category = 'Smartphones'
GROUP BY age_group, brand, category
ORDER BY age_group, brand, category) AS apple_subquery
WHERE brand = 'Apple';

 * mysql+pymysql://root:***@localhost/electronic_sales
1 rows affected.


[(Decimal('1.2122719735'),)]

With the value of 1.21 output, we can conclude that Apple has the ability to increase sales by at least 21% in this category if they market the same as the biggest companies, as the data shows that their products have high enough satisfaction ratings for them to increase their sales. 